subsys/mgmt/mcumgr/mgmt/src/mgmt.c

```diff
@@ -130,9 +135,29 @@ smp_translate_error_fn mgmt_find_error_translation_function(uint16_t group_id)
 void
 mgmt_register_group(struct mgmt_group *group)
 {
+	num_groups++;
 	sys_slist_append(&mgmt_group_list, &group->node);
 }
 
+int mgmt_get_num_groups(void)
+{
+	return num_groups;
+}
+
+int mgmt_get_group_ids(int *ids)
+{
+	int count = 0;
+	sys_snode_t *snp, *sns;
+
+	SYS_SLIST_FOR_EACH_NODE_SAFE(&mgmt_group_list, snp, sns) {
+		struct mgmt_group *loop_group =
+			CONTAINER_OF(snp, struct mgmt_group, node);
+		ids[count] = loop_group->mg_group_id;
+		count++;
+	}
+	return count;
+}
```

Comment:
This should call a callback instead so that the memory usage is bounded not unbounded which is the zephyr way of doing this

Reply:
I'm not fully understanding your comment here. I get that using a callback instead of a direct function call could be more of the zephyr way, but how does it change whether the memory usage is bounded or unbounded? Either way, to get all group IDs, we need to iterate through the whole mgmt_group_list array.

Reply:
If you have 3 groups, this returns 3 and your code creates a stack variable of size 3*4 = 12 bytes. If you have 200 groups, this returns 200 and your code creates a stack variable of size 200*4 = 800 bytes, thus it is unbounded.

Reply:
I understand that, but I'm not clear how moving to a callback implementation would cause the memory to be bounded. If we want a list of all group ids, no matter how we retrieve it, that will inherently be unbounded by the number of groups in the system.

Reply:
So, if I'm understanding correctly, I need to introduce a variable that caps the number of groups that can be returned, and if there are more groups registered than the variable allows, this command will just return an error?

Reply:
There are still errors to work out, but do you think that my updated approach is acceptable? Or, am I at least getting closer?